### PR TITLE
[DEV-2013] V4 updates

### PR DESCRIPTION
**Pull request by jonmarkgo**
See https://github.com/jonmarkgo/omniauth-mlh/pull/1 for comments/description

### CODE REVIEW

File: lib/omniauth/strategies/mlh.rb [hunk 1]
```diff
@@ -5,42 +5,103 @@
...
+                     last_name: data[:last_name],
+                     email: data[:email],
+                     phone_number: data[:phone_number],
+                     roles: data[:roles],
+
+                     # Expandable fields
+                     profile: data[:profile],
+                     address: data[:address],
+                     social_profiles: data[:social_profiles],
+                     professional_experience: data[:professional_experience],
```
Comment (RashikaKarki) [inline lib/omniauth/strategies/mlh.rb]: Now that I am looking at it, we should probably change it to `professional_experiences`.

File: lib/omniauth/strategies/mlh.rb [hunk 2]
```diff
@@ -5,42 +5,103 @@
...
+      end
+
+      private
+
+      def fetch_and_process_data
+        response = access_token.get(build_api_url)
+        data = JSON.parse(response.body, symbolize_names: true)
+        return {} unless data.is_a?(Hash)
+
+        symbolize_nested_arrays(data)
```
Comment (RashikaKarki) [inline lib/omniauth/strategies/mlh.rb:79]: We should be able to use `deep_symbolize_keys`: https://apidock.com/rails/v4.0.2/Hash/deep_symbolize_keys

File: lib/omniauth/strategies/mlh.rb [hunk 3]
```diff
@@ -5,42 +5,103 @@
...
+      def build_api_url
+        url = 'https://api.mlh.com/v4/users/me'
+        expand_fields = options[:expand_fields] || []
+        return url if expand_fields.empty?
+
+        expand_query = expand_fields.map { |f| "expand[]=#{f}" }.join('&')
+        "#{url}?#{expand_query}"
+      end
+
+      def prune_hash(hash)
```
Comment (RashikaKarki) [inline lib/omniauth/strategies/mlh.rb]: We can use `compact`.
https://apidock.com/ruby/Array/compact

File: spec/omni_auth/strategies/mlh_spec.rb [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        },
+        'education' => nil,
+        'social_profiles' => {
+          'github' => {},
+          'linkedin' => {}
+        }
+      }
+
+      it 'handles nil values correctly' do
+        result = strategy.data
```
Comment (RashikaKarki) [inline spec/omni_auth/strategies/mlh_spec.rb:130]: We usually add a line break before the expect block for better readability.

File: spec/omni_auth/strategies/mlh_spec.rb [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        expect(result[:last_name]).to be_nil
+        expect(result[:profile][:gender]).to be_nil
+        expect(result[:profile][:location]).to eq({ city: nil, country: 'USA' })
+        expect(result[:education]).to be_nil
+      end
+
+      it 'handles empty hash structures' do
+        result = strategy.data
+        expect(result[:social_profiles]).to eq({ github: {}, linkedin: {} })
```
Comment (RashikaKarki) [inline spec/omni_auth/strategies/mlh_spec.rb:141]: \n

File: spec/omni_auth/strategies/mlh_spec.rb [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+      end
+    end
+
+    context 'with API error' do
+      it 'returns empty hash on error' do
+        allow(access_token).to receive(:get).and_raise(StandardError)
+        expect(strategy.data).to eq({})
```
Comment (RashikaKarki) [inline spec/omni_auth/strategies/mlh_spec.rb:149]: \n after the stub.

File: spec/omni_auth/strategies/mlh_spec.rb [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+  describe '#uid' do
+    context 'with valid data' do
+      before do
+        allow(strategy).to receive(:data).and_return({ id: 'test-123' })
+      end
+
+      it 'returns the id from the data hash' do
+        expect(strategy.uid).to eq('test-123')
+      end
```
Comment (RashikaKarki) [inline spec/omni_auth/strategies/mlh_spec.rb]: Since, we just have 1 test within the context, we can add the stub within the it block.

File: spec/omni_auth/strategies/mlh_spec.rb [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+    end
+
+    context 'with missing id' do
+      before do
+        allow(strategy).to receive(:data).and_return({})
+      end
+
+      it 'returns nil when id is not present' do
+        expect(strategy.uid).to be_nil
+      end
```
Comment (RashikaKarki) [inline spec/omni_auth/strategies/mlh_spec.rb]: Same as above, stub within the it block.

File: spec/spec_helper.rb [hunk 9]
```diff
@@ -1,23 +1,31 @@
...
-  config.include Rack::Test::Methods
-  config.extend  OmniAuth::Test::StrategyMacros, type: :strategy
-
-  OmniAuth.config.test_mode = true
+  config.expect_with :rspec do |c|
+    c.syntax = :expect
+  end
 end
+
+WebMock.disable_net_connect!(allow_localhost: true)
```
Comment (RashikaKarki) [inline spec/spec_helper.rb:31]: Just curious, why did we allow `localhost`?

File: spec/support/shared_examples.rb [hunk 10]
```diff
@@ -1,46 +1,84 @@
...
-shared_examples 'an oauth2 strategy' do
-  describe '#client' do
-    it 'is initialized with symbolized client_options' do
-      @options = { client_options: { 'authorize_url' => 'https://example.com' } }
+RSpec.shared_examples 'basic data retrieval tests' do
+  context 'with successful API response' do
+    before do
+      allow(access_token).to receive(:get)
+        .with('https://api.mlh.com/v4/users/me')
+        .and_return(instance_double(OAuth2::Response, parsed: { 'data' => base_user_data }))
```
Comment (RashikaKarki) [inline spec/support/shared_examples.rb:8]: The alignment seems to be a bit different than the pattern we use, we should probably use MLH rubocop: https://github.com/MLH/MLH-Rubocop-Config

This probably should be handled in another PR though.

**Review by RashikaKarki — CHANGES_REQUESTED** [review]
Few small changes.

File: lib/omniauth/strategies/mlh.rb (hunk 2 above)
Comment (jonmarkgo) [reply lib/omniauth/strategies/mlh.rb:79]: omniauth-oauth2 (the gem that omniauth-mlh is built on top of) uses something weird called SnakyHash instead of standard Ruby hashes. It was breaking deep_symbolize_keys so I changed it to this

File: spec/support/shared_examples.rb (hunk 10 above)
Comment (jonmarkgo) [reply spec/support/shared_examples.rb:8]: Yeah, the rubocop linter config we have here passes but we may want another PR to update the whole thing to MLH's new rubocop config

File: lib/omniauth/strategies/mlh.rb (hunk 1 above)
Comment (jonmarkgo) [reply lib/omniauth/strategies/mlh.rb]: It's called professional_experience in the API spec from Core right now so this is accurate but we can consider renaming it in a future issue (though it would have to be done in more places than just omniauth-mlh)

File: lib/omniauth/strategies/mlh.rb (hunk 3 above)
Comment (jonmarkgo) [reply lib/omniauth/strategies/mlh.rb]: Will update

**Review by RashikaKarki — APPROVED** [review]
LGTM. Though we might want to wait until V4 is launched, otherwise OHQ and possibly other platforms will be down.

**Comment by jonmarkgo** [comment]
Agreed, let's merge it on Monday - make sure you specify version 2.0.0 in the new ohq changes
